lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -293,16 +296,64 @@ instance Buildable (PartialTx era) where
         cardanoTxF :: Cardano.Tx era -> Builder
         cardanoTxF tx' = pretty $ pShow tx'
 
+-- | Assumptions about the UTxO which are needed for coin-selection.
+--
+-- In practice turns out to be
+--
+-- @
+--  data UTxOAssumptions
+--      = AllKeyPaymentCredentials
+--      | AllByronKeyPaymentCredentials
+--      | AllScriptPaymentCredentialsFrom ScriptTemplate scriptLookup
+-- @
+--
+-- however representing it as such is inconvenient at the moment.
+data UTxOAssumptions = forall k ktype. UTxOAssumptions
+    { txLayer
+        :: TransactionLayer k ktype SealedTx
+        -- TODO: Replace with smaller and smaller parts of 'TransactionLayer'
+    , inputScriptLookup
+        :: Maybe (W.Address -> (CA.Script KeyHash))
+    , inputScriptTemplate
+        :: Maybe ScriptTemplate
+    , description :: String
```

Comment:
Hmm. The existence of this field makes me wonder if you've considered using a sum-type instead:

For example:
```hs
data UTxOAssumptions
    = AllKeyPaymentCredentials
    | AllScriptPaymentCredentials
    | ...
```

Reply:
<img width="587" alt="Skärmavbild 2023-03-29 kl  12 33 46" src="https://user-images.githubusercontent.com/304423/228507402-9cbf1a4a-2dcf-4246-9828-e9d2f0a6554b.png">

The problem at the moment is that we rely on constructing the `UTxOAssumptions` step by step. In particular we have these ugly parameters to the `balanceTransaction` server handler: https://github.com/input-output-hk/cardano-wallet/blob/4b2d62015ec9c8119b01fcb8898f09f0b03122ec/lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs#L2986-L2987

I think this needs to be resolved before switching to a sum-type.

And I have a feeling the solution might be along the lines of https://input-output-rnd.slack.com/archives/C04T7AEDY4X/p1679422712991169?thread_ts=1678777670.296419&cid=C04T7AEDY4X, as both `ScriptTemplate` and `Address -> Script` can be derived from `WalletState (SharedState n SharedKey)`

Reply:
(Had forgotten to push it, but I did drop the `description` field, for this PR at least)